From 065d6350349bfe1b1749b902adc84542446a47b9 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Thu, 18 Jan 2024 13:16:20 +0100
Subject: [PATCH 0827/1141] media: stm32: dcmipp: correct PIPEDIFF handling
 within dcmipp-isp

A unique stream is getting out of the ISP can go into both the pixelproc
subdev of main pipe and aux pipe. A unique pad on isp subdev can thus be
used for that purpose. Connect the aux pixelproc to the pad[1] of the
isp. Since that pad can have 2 remotes pad, implement a dedicated function
to discover if the link to the aux pixelproc is enabled or not.

Change-Id: Ic84bc19793fc6042fcdcb6d491a242a591177e63
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/352412
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-core.c       |  4 +--
 .../st/stm32/stm32-dcmipp/dcmipp-isp.c        | 25 ++++++++++++++++---
 2 files changed, 23 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c
index 59e678453763..5ecbef7f8ccb 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-core.c
@@ -199,13 +199,13 @@ static const struct dcmipp_ent_link stm32mp25_ent_links[] = {
 			MEDIA_LNK_FL_ENABLED | MEDIA_LNK_FL_IMMUTABLE),
 	DCMIPP_ENT_LINK(ID_MAIN_ISP,	1, ID_MAIN_POSTPROC,  0,
 			MEDIA_LNK_FL_ENABLED | MEDIA_LNK_FL_IMMUTABLE),
-	DCMIPP_ENT_LINK(ID_MAIN_ISP,	2, ID_AUX_POSTPROC,  0, 0),
+	DCMIPP_ENT_LINK(ID_MAIN_ISP,	1, ID_AUX_POSTPROC,  0, 0),
 	DCMIPP_ENT_LINK(ID_MAIN_POSTPROC,	1, ID_MAIN_CAPTURE,  0,
 			MEDIA_LNK_FL_ENABLED | MEDIA_LNK_FL_IMMUTABLE),
 	DCMIPP_ENT_LINK(ID_INPUT,	3, ID_AUX_POSTPROC,  0, 0),
 	DCMIPP_ENT_LINK(ID_AUX_POSTPROC,	1, ID_AUX_CAPTURE,  0,
 			MEDIA_LNK_FL_ENABLED | MEDIA_LNK_FL_IMMUTABLE),
-	DCMIPP_ENT_LINK(ID_MAIN_ISP,	3, ID_ISP_STAT_CAPTURE,  0,
+	DCMIPP_ENT_LINK(ID_MAIN_ISP,	2, ID_ISP_STAT_CAPTURE,  0,
 			MEDIA_LNK_FL_ENABLED | MEDIA_LNK_FL_IMMUTABLE),
 };
 
diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
index 7b63d042d7b6..88c3a8a75fce 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -927,6 +927,24 @@ static int dcmipp_isp_colorconv_user(struct dcmipp_isp_device *isp)
 	return dcmipp_isp_colorconv_set(isp, &ccconf);
 }
 
+static bool dcmipp_isp_is_aux_output_enabled(struct dcmipp_isp_device *isp)
+{
+	struct media_link *link;
+
+	for_each_media_entity_data_link(isp->ved.ent, link) {
+		if (link->source != &isp->ved.pads[1])
+			continue;
+
+		if (!(link->flags & MEDIA_LNK_FL_ENABLED))
+			continue;
+
+		if (!strcmp(link->sink->entity->name, "dcmipp_aux_postproc"))
+			return true;
+	}
+
+	return false;
+}
+
 static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
 {
 	struct dcmipp_isp_device *isp = v4l2_get_subdevdata(sd);
@@ -935,7 +953,7 @@ static int dcmipp_isp_s_stream(struct v4l2_subdev *sd, int enable)
 	mutex_lock(&isp->lock);
 	if (enable) {
 		/* Check if link between ISP & Pipe2 postproc is enabled */
-		if (media_pad_remote_pad_first(&sd->entity.pads[2]))
+		if (dcmipp_isp_is_aux_output_enabled(isp))
 			reg_clear(isp, DCMIPP_P1FSCR, DCMIPP_P1FSCR_PIPEDIFF);
 		else
 			reg_set(isp, DCMIPP_P1FSCR, DCMIPP_P1FSCR_PIPEDIFF);
@@ -1117,12 +1135,11 @@ struct dcmipp_ent_device *dcmipp_isp_ent_init(struct device *dev,
 	ret = dcmipp_ent_sd_register(&isp->ved, &isp->sd,
 				     v4l2_dev,
 				     entity_name,
-				     MEDIA_ENT_F_PROC_VIDEO_PIXEL_FORMATTER, 4,
-				     (const unsigned long[4]) {
+				     MEDIA_ENT_F_PROC_VIDEO_PIXEL_FORMATTER, 3,
+				     (const unsigned long[3]) {
 				     MEDIA_PAD_FL_SINK,
 				     MEDIA_PAD_FL_SOURCE,
 				     MEDIA_PAD_FL_SOURCE,
-				     MEDIA_PAD_FL_SOURCE,
 				     },
 				     &dcmipp_isp_int_ops, &dcmipp_isp_ops,
 				     NULL, dcmipp_isp_irq_thread);
-- 
2.39.2

